metadata-suggestions.md

```diff
@@ -0,0 +1,528 @@
+                "place": -1
+              }
+            ]
+        },
+        "type": "metadataSuggestionEntry",
+        "_links": {
+          "self": {
+            "href": "https://dspace7-internal.atmire.com/server/api/integration/metadata-suggestions/orcid/entryValues/0000-0003-3681-2038"
+          },
+         "changes": {
+           "href": "https://dspace7-internal.atmire.com/server/api/integration/metadata-suggestions/orcid/entryValues/0000-0003-3681-2038/changes"
+         }
+        }
+      }
+    ]
+  }
+}
+```
+
+The changes are explained below, and won't be embedded in this endpoint.
+
+### Single entry
+**/api/integration/metadata-suggestions/<:suggestion-name>/entryValues/<:entry-id>**  
+
+Parameters:
+* workspaceitem: the current workspace item ID (mutually exclusive with workflowitem)
+* workflowitem: the current workflow item ID (mutually exclusive with workspaceitem)
+
+It returns the data from one entry in an external source
+
+sample for an external source /api/integration/metadata-suggestions/orcid/entryValues/0000-0002-4271-0436
+```json
+{
+  "id": "0000-0002-4271-0436",
+  "display": "Smith, Dean",
+  "value": "Smith, Dean",
+  "type": "metadataSuggestionEntry",
+  "metadataSuggestion": "orcid",
+  "metadata": {
+    "dc.identifier.orcid": [
+      {
+        "value": "0000-0002-4271-0436",
+        "language": null,
+        "authority": null,
+        "confidence": 0,
+        "place": -1
+      }
+    ],
+    "dc.identifier.uri": [
+      {
+        "value": "https://orcid.org/0000-0002-4271-0436",
+        "language": null,
+        "authority": null,
+        "confidence": 0,
+        "place": -1
+      }
+    ],
+    "person.familyName": [
+      {
+        "value": "Smith",
+        "language": null,
+        "authority": null,
+        "confidence": 0,
+        "place": -1
+      }
+    ],
+    "person.givenName": [
+      {
+        "value": "Dean",
+        "language": null,
+        "authority": null,
+        "confidence": 0,
+        "place": -1
+      }
+    ]
+  },
+  "_links": {
+    "self": {
+      "href": "https://dspace7-internal.atmire.com/server/api/integration/metadata-suggestions/orcid/entryValues/0000-0002-4271-0436"
+    },
+    "changes": {
+      "href": "https://dspace7-internal.atmire.com/server/api/integration/metadata-suggestions/orcid/entryValues/0000-0002-4271-0436/changes"
+    }
+  },
+  "_embedded" : {
+     "changes" : [
+      {
+        "op": "add",
+        "path": "/metadata/dc.identifier.orcid",
+        "value": [ { "value": "0000-0002-4271-0436" } ]
+      },
+      {
+        "op": "add",
+        "path": "/metadata/dc.identifier.uri/-",
+        "value": [ { "value": "https://orcid.org/0000-0002-4271-0436" } ]
+      }
+     ]
+  }
+}
+```
+
+The changes embedded are the suggested metadata changes to be applied to the given item
+
+### Changes for a single entry
+**/api/integration/metadata-suggestions/<:suggestion-name>/entryValues/<:entry-id>/changes**  
```

Comment:
This endpoint seems odd to me too... if you have the `workspaceitem` or `workflowitem` passed into `/api/integration/metadata-suggestions/<:suggestion-name>/entryValues/<:entry-id>`, couldn't you just return the "changes" from that endpoint?  I'm not sure this `changes` subpath is necessary, plus it doesn't seem to follow our REST Best Practices.  It's not really a "subresource" of the `entry` (represented by `entry-id`).  It's more like a different "view" or "filter" of that entry.  I'm not sure what to suggest here though yet, as I'm still trying to understand the parent endpoints here.

Reply:
The `changes` are currently assumed not to be embedded in the `/api/integration/metadata-suggestions/<:suggestion-name>/entries` endpoint for performance reasons (although it's hard to estimate the performance impact prior to the implementation).
In order to ensure this scales well, it was introduced as a sub-resource. The changes are calculated based on the combination of the information in the entry and in the item

Reply:
@benbosman : To clarify, my original statement here, I wasn't suggesting to embed `changes`...instead, I was suggesting that this specific endpoint _might_ return a different response based on parameters.  So for example:

* ` GET /api/integration/metadata-suggestions/<:suggestion-name>/entryValues/<:entry-id>` would return the full external entry (which would seem to be nearly identical to what  `/api/integration/externalsources/<:source-name>/entryValue/<:entry-id>` might return (if the backends were identical)).
*  `GET /api/integration/metadata-suggestions/<:suggestion-name>/entryValues/<:entry-id>?workspaceitem=1111-2222-3333` would return the difference (i.e. the "changes") between `<:entry-id>` and that `WorkSpaceItem`

So, this would mean you'd no longer need a subresource called `changes` (which as I said, doesn't seem like a normal subresource anyways).

It's just a different idea of how to possibly approach this specific endpoint.

Reply:
Since the features of both endpoints are quite different, it doesn't make much sense to ask for the metadata of e.g. PubMed's live import using /api/integration/externalsources/pubmed/entryValue/31680121 especially since there's a reasonable chance /api/integration/externalsources/pubmed doesn't exist (is not configured)

It would be similar to updating the collection metadata using /api/core/items/<:collectionID> just because items also have metadata, even though there's no item with that ID

Reply:
This `changes` subresource still bugs me, as it doesn't seem very RESTful.  I understand the purpose here, and it seems like the goal is to simplify the client side processing, so the client side doesn't need to compare the response of the `/metadata-suggestions` request to the current metadata in the `WorkspaceItem/WorkflowItem`.  

However, at the same time, I'm not sure the Server Side (REST API) will always **know** whether the `/changes` should return an `add` or `replace` or `delete` operation.  For example, if the `/metadata-suggestion` returns one `dc.title` while the WorkspaceItem has a different `dc.title`, is that an `add` operation (to end up with two alternative titles) or a `replace` operation (to end up with one title)?  Would the Server Side be able to make a proper `/changes` suggestion without input from an End User?

I'm starting to wonder if it's easier to drop this `/changes` endpoint altogether, and do the comparison on the Client Side (as these are simple string comparisons).  In the above scenario, I could imagine a UI that simply shows that there's a difference in the values of `dc.title` to let the End User select the one they prefer, or even allow them to select both.  Based on the User's selection, the UI would generate either a noop, an `add` or a `replace` using the existing Item PATCH metadata functionality.

This is also a topic that we could bring to the DSpace 7 Working Group meeting -- as this feature is not specific to Entities in any way.  Maybe others see a different way of achieving this endpoint in a RESTful fashion.